docs/examples/preprocess.py

```diff
@@ -31,36 +86,170 @@ def preprocess():
+        if filename.startswith("examples/"):
+            # all good, the path is absolute already.
+            return DOCS_ROOT / filename
+        elif "/" not in filename:
+            # Path is like `job.sh.diff`, which is good.
+            return example_readme_template_path.parent / filename
+        else:
+            raise NotImplementedError(
+                f"Example {example_readme_template_path} uses a weird format in a "
+                f"literalinclude: {filename}"
+            )
+
+    new_content_lines: list[str] = (
+        textwrap.dedent(
+            f"""\
+            .. NOTE: This file is auto-generated from {relative_readme_template_path}
+            .. This is done so this file can be easily viewed from the GitHub UI.
+            .. **DO NOT EDIT**
+            """
+        ).splitlines()
+        + [""]  # Need an empty line between this header and the content, otherwise rstcheck cries
+    )
+
+    # Matches a group of non-whitespace characters after the literalinclude directive, maybe
+    # followed by some whitespace.
+    literalinclude_pattern = r".. literalinclude::\s*(?P<file_path>[^\s]+)\s*"
+
+    # Matches some leading whitespace since it needs to be idented, (note: the number of spaces
+    # varies, usually 3 or 4), followed by a word and potentially some whitespace at the end.
+    language_pattern = r"\s+:language:\s*(?P<language>[^\s]+)\s*"
+
+    line_index = 0
+    # Using a while loop here because we want to skip two lines whenever we find a match.
+    while line_index < len(content_lines):
+        line = content_lines[line_index]
+        # Using fullmatch with the spaces specified in the regex so that we don't match comments.
+        # There is also no need to call .strip() or the result this way.
+        include_block_match = re.fullmatch(literalinclude_pattern, line)
+        if not include_block_match:
+            # This is just a normal text line in the doc. Add it and move to the next.
+            new_content_lines.append(line)
+            line_index += 1
+            continue
+
+        # Can't have a literalinclude on the last line
+        assert line_index + 1 < len(content_lines)
+        language_match = re.fullmatch(language_pattern, content_lines[line_index + 1])
+        if not language_match:
+            # NOTE: Add 1 to line index so paths in logs are clickable for debugging purposes.
+            raise RuntimeError(
+                f"Found a literalinclude block at "
+                f"{example_readme_template_path}:{line_index+1} but it's missing a "
+                f":language: directive on the following line."
+            )
+
+        def _get_block_length() -> int:
+            length = 1
+            while (
+                line_index + length < len(content_lines)
+                and content_lines[line_index + length]
+                and not content_lines[line_index + length].isspace()
+            ):
+                length += 1
+            return length
```

Comment:
I haven't tested so maybe I'm following the code wrong, but you only stop on empty/whitespace lines, so it seems to me that the function will return 4 on the following block:

```rst
.. literalinclude:: x.py
  :language: python
.. important::
  Don't run vscode on the login nodes you guys.
```

And 1 on this one, which I'm not certain is legal or not:

```rst
.. literalinclude:: x.py

  :language: python
```

Reply:
Fixed here: 806c59da5ecc90e3393aead927628820d90ea94d